Watcher errors in combination with concurrent playback

**Describe the bug**
Except for on startup, the watcher always assumes that the filesystem structure represents its internal tree representation and the state of the ZODB at the transaction `self.last_visible_txn`. However, if a `pick`, `playback` etc. is used, this might not be the case. For example, if `playback` is used to delete an object, the watcher will terminate because it tries to move the no longer existing folder into `__orphans__` before deleting it.
It does this because the transaction might consist of a move, in which case another object might find itself the new parent of the object, so we need to fetch it again from `__orphans__`. This will currently never happen for transactions that were created by `zodbsync` itself by playback etc., because it does not recognize a move, only deletions, changes and new objects. But the watcher does not know that this transaction will not contain any moves and it might have to deal with a couple of transactions at once that it missed since its last sweep.

**To Reproduce**
Steps to reproduce the behavior:
1. Start the watcher on a popuplated repository
2. Delete a folder inside the repository
3. Run playback on the folder to propagate the deletion into the ZODB
4. Wait for the watcher to crash

**Expected behavior**
Either the watcher needs to be able to deal with the situation that the filesystem might sometimes look like the ZODB at transaction B instead of A (A being the transaction from the last watcher step and B from the current one) or any command that works in the direction filesystem -> ZODB should somehow signal to the watcher that the filesystem now corresponds to transaction B. It might write into `__last_txn__`, and the watcher might read it. Since the `playback` command can not really be certain that the filesystem is actually up to date at state `B` (there might me other changes to unrelated objects or someone might only play back some of the changes that they do on the file system), the first possibility is probably to be preferred.

This might be related to sporadic failures in the watcher tests, like:
```
____________________________________________ TestSync.test_watch_change _____________________________________________

self = <perfact.zodbsync.tests.test_sync.TestSync object at 0x7f2adc59af10>
conn = <perfact.zodbsync.helpers.Namespace object at 0x7f2adc4c5190>

    def test_watch_change(self, conn):
        """
        Start the watcher, change something using the second connection without
        commiting yet, do a step on the watcher, make sure the change is not
        yet visible, then commit the change and do another step, making sure
        that it is now present.
        """
        fname = self.repo.path + '/__root__/__meta__'
        watcher = self.runner('watch')
        watcher.setup()
        conn.tm.begin()
        conn.app._addRole('TestRole')
        watcher.step()
        assert 'TestRole' not in open(fname).read()
        conn.tm.commit()
        watcher.step()
>       assert 'TestRole' in open(fname).read()
E       assert 'TestRole' in "[\n    ('is_root', True),\n    ('perms', [\n        ('Access contents information', False, ['Anonymous', 'Manager']),...bDAV access', False, ['Authenticated', 'Manager']),\n        ]),\n    ('title', 'Zope'),\n    ('type', 'Folder'),\n]\n"
E        +  where "[\n    ('is_root', True),\n    ('perms', [\n        ('Access contents information', False, ['Anonymous', 'Manager']),...bDAV access', False, ['Authenticated', 'Manager']),\n        ]),\n    ('title', 'Zope'),\n    ('type', 'Folder'),\n]\n" = <built-in method read of _io.TextIOWrapper object at 0x7f2add70cc70>()
E        +    where <built-in method read of _io.TextIOWrapper object at 0x7f2add70cc70> = <_io.TextIOWrapper name='/tmp/tmppxm02pj_/__root__/__meta__' mode='r' encoding='UTF-8'>.read
E        +      where <_io.TextIOWrapper name='/tmp/tmppxm02pj_/__root__/__meta__' mode='r' encoding='UTF-8'> = open('/tmp/tmppxm02pj_/__root__/__meta__')

perfact/zodbsync/tests/test_sync.py:356: AssertionError
```
or
```
_____________________________________________ TestSync.test_watch_move ______________________________________________

self = <perfact.zodbsync.tests.test_sync.TestSync object at 0x7fc23c38e100>
conn = <perfact.zodbsync.helpers.Namespace object at 0x7fc23c36d880>

    def test_watch_move(self, conn):
        """
        Create a Page Template, record it using the watcher, rename it and make
        sure the watcher notices. Then add a second one and do a
        three-way-rename in one transaction, making sure the watcher keeps
        track.
        """
        watcher = self.runner('watch')
        watcher.setup()
        root = self.repo.path + '/__root__/'
        src = '/__source-utf8__.html'
        app = conn.app
    
        add = app.manage_addProduct['PageTemplates'].manage_addPageTemplate
        rename = app.manage_renameObject
    
        with conn.tm:
            add(id='test1', text='test1')
        watcher.step()
        import pdb
        pdb.set_trace()
>       assert os.path.isdir(root + 'test1')
E       AssertionError: assert 'test2' == 'test1'
E         - test1
E         ?     ^
E         + test2
E         ?     ^

perfact/zodbsync/tests/test_sync.py:365: AssertionError
```

--- a/perfact/zodbsync/commands/watch.py
+++ b/perfact/zodbsync/commands/watch.py
@@ -494,22 +494,29 @@ class Watch(SubCommand):
         changes."""
         self.unregister_signals()
         self.acquire_lock(timeout=300)
-        self.register_signals()
+        try:
+            self.register_signals()
 
-        start_txnid = _increment_txnid(self.last_visible_txn)
-        self._set_last_visible_txn()
-        self._read_changed_oids(
-            txn_start=start_txnid,
-            txn_stop=self.last_visible_txn,
-        )
-        # make sure we see a consistent snapshot, even though we later
-        # abort this transaction since we do not write anything
-        self.sync.tm.begin()
-        self._update_objects()
-        self.sync.tm.abort()
+            start_txnid = _increment_txnid(self.last_visible_txn)
+            self._set_last_visible_txn()
+            self._read_changed_oids(
+                txn_start=start_txnid,
+                txn_stop=self.last_visible_txn,
+            )
+            # make sure we see a consistent snapshot, even though we later
+            # abort this transaction since we do not write anything
+            self.sync.tm.begin()
+            self._update_objects()
+            self.sync.tm.abort()
 
-        self._store_last_visible_txn()
-        self.release_lock()
+            self._store_last_visible_txn()
+        except TreeOutdatedException:
+            self.logger.info(
+                'Exiting due to inconsistencies in filesystem'
+            )
+            self.exit.set()
+        finally:
+            self.release_lock()
 
     def run(self, interval=10):
         """ Setup and run in a loop. """
@@ -520,10 +527,6 @@ class Watch(SubCommand):
         else:
             self.spawned_setup()
         while not self.exit.is_set():
-            try:
-                self.step()
-            except TreeOutdatedException:
-                self.release_lock()
-                self.exit.set()
+            self.step()
             # a wait that is interrupted immediately if exit.set() is called
             self.exit.wait(interval)

--- a/perfact/zodbsync/commands/watch.py
+++ b/perfact/zodbsync/commands/watch.py
@@ -49,6 +49,12 @@ def _increment_txnid(s):
     return bytes(arr)
 
 
+class TreeOutdatedException(Exception):
+    """Exception which is raised if the internal tree structure
+    is not matching the actual Filesystem anymore"""
+    pass
+
+
 class Watch(SubCommand):
     """Periodically check for changes and record them"""
     # Connects to ZEO, builds a mirror of the tree structure of the objects,
@@ -311,7 +317,12 @@ class Watch(SubCommand):
                     )
                 )
                 os.makedirs(self.base_dir+newpath)
-                os.rename(self.base_dir+oldpath, self.base_dir+newpath)
+                try:
+                    os.rename(self.base_dir+oldpath, self.base_dir+newpath)
+                except OSError as err:
+                    if err.errno == 2:  # no such file or directory
+                        raise TreeOutdatedException()
+                    raise
                 self._update_path(child_oid, newpath)
 
         # go through new children and check if they have old parents
@@ -509,6 +520,10 @@ class Watch(SubCommand):
         else:
             self.spawned_setup()
         while not self.exit.is_set():
-            self.step()
+            try:
+                self.step()
+            except TreeOutdatedException:
+                self.release_lock()
+                self.exit.set()
             # a wait that is interrupted immediately if exit.set() is called
             self.exit.wait(interval)

--- a/perfact/zodbsync/commands/watch.py
+++ b/perfact/zodbsync/commands/watch.py
@@ -322,6 +322,7 @@ class Watch(SubCommand):
                 except OSError as err:
                     if err.errno == 2:  # no such file or directory
                         raise TreeOutdatedException()
+                    self.logger.exception('Unexpected OSError')
                     raise
                 self._update_path(child_oid, newpath)
 

--- a/perfact/zodbsync/tests/test_sync.py
+++ b/perfact/zodbsync/tests/test_sync.py
@@ -6,6 +6,7 @@ import six
 import subprocess
 import pickle
 import pytest
+import shutil
 
 import ZEO
 import transaction
@@ -22,6 +23,7 @@ from .. import helpers
 from .. import extedit
 from .. import object_types
 from . import environment as env
+from ..commands import watch
 
 
 class DummyResponse():
@@ -868,6 +870,47 @@ class TestSync():
             meta = f.read()
         self.watcher_step_until(watcher,
                                 lambda: "('title', '"+new_title+"')" in meta)
+
+    def test_watch_structure_changes_and_playback_deleted_folder(self, conn):
+        """
+        create structure while 'watch' command is running,
+        remove a folder, then play those changes back and check,
+        whether the step function correctly "crashes"
+        """
+
+        # start watch daemon
+        watcher = self.mkrunner('watch')
+        watcher.setup()
+        app = conn.app
+        folder_1 = "folder_1"
+        s_folder_1 = "s_folder_1"
+
+        # create folder and wait until watch notices change
+        with conn.tm:
+            app.manage_addFolder(id=folder_1)
+        self.watcher_step_until(watcher,
+                                lambda: os.path.isdir(
+                                    self.repo.path + '/__root__/'+folder_1))
+
+        # create subfolder and wait until watch notices change
+        with conn.tm:
+            app.folder_1.manage_addFolder(id=s_folder_1, title=s_folder_1)
+        path = self.repo.path + '/__root__/'+folder_1+'/'+s_folder_1
+        self.watcher_step_until(watcher,
+                                lambda: os.path.isdir(path))
+
+        # remove folder s_folder_1
+        shutil.rmtree(path)
+
+        # playback changes and check if those are existent in zodb
+        self.run('playback', '/')
+        # assert new_title == self.app.folder_1.s_folder_1.title
+
+        # wait for watch to notices played back changes
+        with pytest.raises(watch.TreeOutdatedException):
+            self.watcher_step_until(watcher, lambda: True)
+        # release the lock the same way the watcher process does
+        watcher.release_lock()
 
     def test_commit_on_branch_and_exec_merge(self):
         '''

--- a/perfact/zodbsync/tests/test_sync.py
+++ b/perfact/zodbsync/tests/test_sync.py
@@ -23,7 +23,6 @@ from .. import helpers
 from .. import extedit
 from .. import object_types
 from . import environment as env
-from ..commands import watch
 
 
 class DummyResponse():
@@ -906,10 +905,7 @@ class TestSync():
         self.run('playback', '/')
 
         # wait for watch to notices played back changes
-        with pytest.raises(watch.TreeOutdatedException):
-            self.watcher_step_until(watcher, lambda: True)
-        # release the lock the same way the watcher process does
-        watcher.release_lock()
+        self.watcher_step_until(watcher, watcher.exit.is_set)
 
     def test_commit_on_branch_and_exec_merge(self):
         '''

--- a/perfact/zodbsync/tests/test_sync.py
+++ b/perfact/zodbsync/tests/test_sync.py
@@ -904,7 +904,6 @@ class TestSync():
 
         # playback changes and check if those are existent in zodb
         self.run('playback', '/')
-        # assert new_title == self.app.folder_1.s_folder_1.title
 
         # wait for watch to notices played back changes
         with pytest.raises(watch.TreeOutdatedException):

